task-tracker-webapp/src/main/java/com/isa/tasktrackerwebapp/controller/RegistrationController.java

```diff
@@ -0,0 +1,89 @@
+import org.springframework.web.bind.annotation.RequestParam;
+
+@Controller
+class RegistrationController {
+    private final LoginService loginService;
+    private final RegistrationService registrationService;
+    private final LoginValidator loginValidator;
+
+    RegistrationController(LoginService loginService, RegistrationService registrationService, LoginValidator loginValidator) {
+        this.loginService = loginService;
+        this.registrationService = registrationService;
+        this.loginValidator = loginValidator;
+    }
+
+    private static final Logger logger = LoggerFactory.getLogger(RegistrationController.class);
+
+    @GetMapping("/registration")
+    String registration(Model model, User user) {
+        model.addAttribute("content", PageType.REGISTRATION.getContentValue())
+                .addAttribute("pageTitle", PageType.REGISTRATION.getTitleValue())
+                .addAttribute("user", user);
+
+        if (loginService.isUserLoggedIn()) {
+            model.addAttribute("isUserLoggedIn", loginService.isUserLoggedIn());
+        }
+        return "main";
+    }
+
+    @PostMapping("/registerUser")
+    String registerUser(@ModelAttribute @Valid User user, BindingResult bindingResult,
+                        @RequestParam("repeatPassword") String repeatPassword,
+                        Model model) {
+        boolean hasErrors = bindingResult.hasErrors();
+        boolean loginExists = loginValidator.doesLoginExist(user.getLogin());
+        boolean emailExists = registrationService.doesEmailExist(user.getEmail());
+        boolean passwordsDontMatch = !repeatPassword.equals(user.getPassword());
+
+        StringBuilder failureReasonsBuilder = new StringBuilder("Registration attempt failed: ");
+
+        if (hasErrors) {
+            failureReasonsBuilder.append(" user fields validations not passed;");
+        }
+
+        if (loginExists) {
+            failureReasonsBuilder.append(" login already exists;");
+        }
+
+        if (emailExists) {
+            failureReasonsBuilder.append(" email already exists;");
+        }
+
+        if (passwordsDontMatch) {
+            failureReasonsBuilder.append(" passwords don't match;");
+        }
+
+        if (hasErrors || loginExists || emailExists || passwordsDontMatch) {
+            logger.warn(failureReasonsBuilder.toString());
+
+            model.addAttribute("content", PageType.REGISTRATION.getContentValue())
+                    .addAttribute("pageTitle", PageType.REGISTRATION.getTitleValue())
+                    .addAttribute("user", user)
+                    .addAttribute("loginExists", loginExists)
+                    .addAttribute("emailExists", emailExists)
+                    .addAttribute("passwordsDontMatch", passwordsDontMatch);
+
+            return "main";
+        }
+
+        registrationService.registerNewUser(user);
+        logger.info("Registered new user: {}", user);
+        return "redirect:/login?registrationSuccessful";
```

Comment:
co to za zapis? jak dodajemy parametry do urla to właśnie poprzez `?`, ale po nazwie parametru wypadałoby dodać jeszcze jego wartość